Copyright and attribution (#268)
open-energy-transition/handbook: docs/Engineering/CopyrightAndAttribution.mdx

---
sidebar_label: "Copyright and attribution"
---

import MaintainerTag from "@site/src/components/MaintainerTag";
import ControlledDocBanner from "@site/src/components/ControlledDocBanner";

<MaintainerTag maintainerEmails={["thomas.gilon@openenergytransition.org", "johannes.hampp@openenergytransition.org"]} />

# Copyright and attribution

This page describes how we
* handle copyright notices, and 
* attribution of work

> in our work, our own repositories and soft forks of other projects.

Reviewer: ```suggestion
This page describes how we handle copyright notices and attribution in our work, our own repositories and soft forks of other projects.
```

> It provides generic guidelines for licensing and attributing both for original work and the work done in a specific project.

Reviewer: ```suggestion
It provides generic guidelines for licensing and attributing both the original work and the work done in a specific project.
```

> This page builds on the [Soft fork strategy](SoftForkStrategy) and covers also projects that are forked from another project (designated as `upstream` in the following) which is not owned by Open Energy Transition (OET).

Reviewer: ```suggestion
This page builds on the [Soft fork strategy](SoftForkStrategy) and covers projects that are forked from another project (referred to as `upstream` below), which is not owned by Open Energy Transition (OET).
```

A soft fork extends the original work in various ways, including custom code, documentation, data, and configuration.
This page assumes you already have a fully functional soft fork from another project.
For illustrative purposes, this page uses `PyPSA-Eur` as the upstream project and `Open-TYNDP` as the soft-fork project.

## About copyright

> Our work that we create as OET employees is usually protected by copyright law.

Reviewer: ```suggestion
As OET employees, the work that we create is usually protected by copyright law.
```

> This means that code and documentation that we write, or data and databases that we curate and create, can be protected through copyright law.

Reviewer: ```suggestion
This means that the code and the documentation that we write, and the data and the databases that we curate and create, can be protected by copyright law.
```

> In the context of copyright law, the term *work* is usually used to refer the code, documentation, data, databases, ... .

Reviewer: ```suggestion
In this context, the term *work* usually refers to code, documentation, data, databases, among others things.
```

> Copyright laws are different in each country and may differ significantly.

Reviewer: ```suggestion
Copyright laws differ significantly from country to country.
```

> As a non-profit organisation that operates internationally, we need to keep this in mind when discussing details,

Reviewer: ```suggestion
As a international non-profit organisation, we need to keep this in mind when discussing details,
```

> because not all *work* may always be protectable by copyright law in all countries.

Reviewer: ```suggestion
because not all *work* may always be protected by copyright law in every country.
```

Author: ```suggestion
because not all *work* may always be protectable under copyright law in all countries.
```

Broadly speaking in simplified terms, copyright distinguishes between

* *the author*, who created the *work*, and
* *the copyright holder*, who controls usage and distribution of the *work*.

The difference can be understood as *the author* owning the *moral rights* to the *work*,

> while *the copyright holder* owns the *economic rights* to the *work*.

Reviewer: ```suggestion
while *the copyright holder* owns the *economic rights*.
```

:::info

    As an OET employee, you are the *author* of your work and retain the *moral rights* for the *work* you create.

>     Because of your contract with OET, OET is the *copyright holder* of your work and holds the *economic rights* to your work.

Reviewer: ```suggestion
    Because of your contract with OET, OET is the *copyright holder* of your work and holds the *economic rights* to it.
```

:::

> When working in a project with multiple contributors, the *work* is usually created by multiple authors.

Reviewer: ```suggestion
When working on a project with multiple contributors, the *work* is usually created by several authors.
```

If all authors are OET employees, OET is the sole copyright holder of the *work*.

> If some authors are not OET employees, their *work* is likely copyrighted by their employer or by themselves.

Reviewer: ```suggestion
If some authors are not OET employees, their *work* is likely to be copyrighted by their employer or by themselves.
```

Author: ```suggestion
If some authors are not OET employees, they themselves or their employer usually are the copyright holder of their *work*.
```

We usually license the work as a whole, rather than individual contributions, and use the term *contributors* to refer to all authors of the *work*.

> ### Differences between OET projects and contributions to other projects

Reviewer: This could be the last paragraph of the intro. It feels weird to have it here as the only section of this level.

Author: I've adjusted it, see if you like it better that way

> Because of how copyright works, there is a difference between how we handle our own projects and how we handle contributions to other projects.

Reviewer: ```suggestion
Because of how copyright works, we handle our own projects differently from contributions to other projects.
```

> First we discuss how we handle copyright and attribution in our own projects, including new and stand-alone projects, as well as soft forks of other projects.

Reviewer: ```suggestion
First, we discuss how copyright and attribution are handled in our own projects, including new and stand-alone projects, as well as soft forks of other projects.
```

> At the end of this page, we discuss how we handle contributions to other projects, including upstream projects.

Reviewer: ```suggestion
Then, we discuss contributions to other projects, including upstream projects.
```

## OET projects

:::note

    This section describes how we handle copyright and attribution in our own projects, including new and stand-alone projects, as well as soft forks of other projects.
    It does not cover other aspects of a soft fork's README, documentation, or GitHub repository.

:::

> There are mutliple components that work hand-in-hand to ensure that our projects correctly attribute the works and are properly licensed.

Reviewer: ```suggestion
Multiple components work together to ensure that our projects correctly attribute the works and are properly licensed.
```

### README

The README file is often the first document users will read about your project.

> This document is key to listing the copyright information and license of a project.

Reviewer: ```suggestion
This document is essential for providing the copyright information and the license of a project.
```

If you are working with a soft fork of another project, the README is also the place to explain how your project relates to the upstream project.

> 

Reviewer: ```suggestion

#### Generic guidelines

```

> From the licensing and attribution perspective, here are some recommended elements for your README:

Reviewer: ```suggestion
From the licensing and attribution perspective, here are some recommended modifications for your README:
```

> * A *Contributors* section listing the contributors to the project for smaller projects listing all contributors to the project.

Reviewer: ```suggestion
* Add a *Contributors* section listing all contributors to the project, for smaller projects.
```

>   For larger projects this section can be moved into the documentation.

Reviewer: ```suggestion
  For larger projects, this section can be moved to the documentation.
```

> * A *License* section listing the license of the project and the copyright holders.

Reviewer: ```suggestion
* Add a *License* section providing the license and the copyright holders of the project.
  If your project uses multiple licenses in the same repository, e.g. for code, data and metadata, describe the licensing strategy in this section.
```

>   For larger projects, the copyright holders may be abridged as "The contributors to \<project>" and the section then refer to a longer list of contributors in the documentation.

Reviewer: ```suggestion
  For larger projects, the copyright holders may be abridged as "Contributors to \<project>". This section must then refer to a longer list of contributors in the documentation.
```

Author: Actually... what do you think about `<project> contributors`?
That'd be in line with others, like "Wikipedia contributors", "OpenStreetMap contributors",

Reviewer: I kept the PyPSA DNA, here are some examples: [pypsa-eur](https://github.com/PyPSA/pypsa-eur/blob/master/scripts/add_brownfield.py#L1), [atlite](https://github.com/PyPSA/atlite/blob/master/atlite/aggregate.py#L1) and [powerplantmatching](https://github.com/PyPSA/powerplantmatching/blob/master/powerplantmatching/cleaning.py#L1).

I always prefer consistency. It's easier to remember.

>   If your project uses multiple licenses in the same repository, e.g. for code, data and metadata, describe the licensing strategy in this section.

Reviewer: ```suggestion
```

> * A *badge* showing the license of the project, e.g. [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/license/mit/).

Reviewer: ```suggestion
* Add a *badge* showing the license of the project, e.g. [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/license/mit/).
```

  
  ```md
  [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/license/mit/)
  ```

> * If your project uses `REUSE` (see below), add a note that `REUSE` is used to manage the licensing of the project, e.g.

Reviewer: ```suggestion
* If your project uses `REUSE` (see [REUSE tool](#reuse-tool)), add a note stating that `REUSE` is used to manage the licensing of the project, e.g.
```

    
    ```md
    This project uses [REUSE](https://reuse.software/) to manage its licensing.
    ```
  

>   and add a badge to show the compliance with REUSE, e.g. [![REUSE status](https://api.reuse.software/badge/github.com/open-energy-transition/open-tyndp)](https://api.reuse.software/info/github.com/open-energy-transition/open-tyndp).

Reviewer: ```suggestion
  Add a badge to show the compliance with REUSE, e.g. [![REUSE status](https://api.reuse.software/badge/github.com/open-energy-transition/open-tyndp)](https://api.reuse.software/info/github.com/open-energy-transition/open-tyndp).
```

  
  ```md
  [![REUSE status](https://api.reuse.software/badge/github.com/open-energy-transition/open-tyndp)](https://api.reuse.software/info/github.com/open-energy-transition/open-tyndp)
  ```
  

>   The badge requires a one-time registration of the repository with REUSE [here](https://api.reuse.software/register).

Reviewer: ```suggestion
  The badge requires a one-time registration of the repository on REUSE [website](https://api.reuse.software/register).
```

Author: ```suggestion
  The badge requires a one-time registration of the repository on  the [REUSE API website](https://api.reuse.software/register).
```

> 

Reviewer: ```suggestion

#### Soft fork specific guidelines

```

Author: Nothing to change here?

Reviewer: I suggest adding a header.

> If you have a soft fork, make sure you to adjust the following parts in the README:

Reviewer: ```suggestion
If you have a soft fork, here are some additional recommendations:
```

* Explain how to report issues and clearly differentiate between upstream and project-specific issues.
* Update the contact information.

> * When you provide a short summary of your licensing strategy.

Reviewer 2: ```suggestion
* When you provide a short summary of your licensing strategy:
```

Reviewer 2: Or did you intend the following lines to be a sub-list?

Author: A sublist - the list somehow got lost.

* Explain how to report issues and clearly differentiate between upstream and project-specific issues.
* Update the contact information.
* When you provide a short summary of your licensing strategy.
    Describe whether it follows the same strategy as the *upstream* project or not, and explain how it differs from the upstream strategy.
    Clearly indicate that this project is based on the upstream project and link to the licensing information of the upstream project.
    Refer to the detailed version of your licensing strategy in your documentation (see also [Documentation](#documentation) below).

> * Use a badge to show how close your soft fork is to the upstream version (e.g., [demo badge](https://img.shields.io/github/commits-difference/open-energy-transition/open-tyndp?base=master&head=PyPSA:pypsa-eur:master&label=commits%20behind%20PyPSA-Eur) from Open-TYNDP).

Reviewer: ```suggestion
* Use a badge to show how close your soft fork is to the upstream version (e.g., ![demo badge](https://img.shields.io/github/commits-difference/open-energy-transition/open-tyndp?base=master&head=PyPSA:pypsa-eur:master&label=commits%20behind%20PyPSA-Eur)).
```

    # Generic
    ![Commits behind [<upstream project>](https://img.shields.io/github/commits-difference/<organisation>]/<repository>?base=<branch>&head=<upstream organisation>:<upstream repository>:<upstream branch>&label=commits%20behind%20<upstream project>)

    # Example for Open-TYNDP (soft fork of PyPSA-Eur)

>     ![Commits behind PyPSA-Eur](https://img.shields.io/github/commits-difference/open-energy-transition/open-tyndp?base=master&head=PyPSA:pypsa-eur:master&label=commits%20behind%20PyPSA-Eur) 

Reviewer 2: Isn't it also relevant / perhaps more interesting to see how many commits ahead this project is compared to upstream?

Author: I think with this badge the idea was to show how "up to date" the soft fork is with it's origin.
If the fork is for a project, it will always be commits ahead (project specific commits).
But it can be up-to-date with origin for general model improvements.

Reviewer: Correct, that was my motivation behind. Of course, a project using a soft-fork will have ahead commits. However, there is no guarantee at all that the project will be kept up to date to upstream, hence the badge.

>     ![Commits behind [<upstream project>](https://img.shields.io/github/commits-difference/<organisation>]/<repository>?base=<branch>&head=<upstream organisation>:<upstream repository>:<upstream branch>&label=commits%20behind%20<upstream project>)

Reviewer: This is inconsistent with the later use of `[CONTRIBUTORS]` in the REUSE section.

Author: What do you mean exactly?

Reviewer: Variables in code are currently formatted in two ways: `[SOMETHING]` and `<SOMETHING>`. I cannot find a satisfactory solution since both syntaxes (`[...]` and `<...>`) are necessary.

    ```

:::tip

    To see how these recommendations can be applied to a README, you can refer to the [Open-TYNDP README](https://github.com/open-energy-transition/open-tyndp/blob/master/README.md).

:::

### Documentation

> The project documentation is equally important to the README and requires careful revision.

Reviewer 2: ```suggestion
The project documentation is equally important as the README and requires careful revision.
```
Or even `is as important as the`

#### Generic guidelines

> A proper documentation should contain the following sections:

Reviewer 2: ```suggestion
Proper documentation should contain the following sections:
```

> A proper documentation should contain the following sections:

Reviewer: ```suggestion
The project documentation is equally important to the README and requires careful revision.

#### Generic guidelines

A proper documentation should contain the following sections:
```

> * A *license* section with the licensing terms of the project and the copyright holders.

Reviewer: ```suggestion
* An up-to-date *license* section with the licensing terms of the project and the copyright holders.
```

>   This can be the same version as in the README or a more detailed version.

Reviewer: ```suggestion
  This could be the same version as the one in the README, or a more detailed.
```

  Note that any redundant information given here and in the README must be kept in sync with the reality in the project, so try to avoid redundancy.

>   E.g. by linking to only a single list of contributors.

Reviewer: ```suggestion
  Note that any redundancy with the README must be kept in sync, so try to avoid redundancy (e.g. by using a single list of contributors).
```

Author: ```suggestion
  Note that any redundancy with the README must be kept in sync, so try to avoid redundancy (e.g. by using a single list of contributors to link to).
```

  For simplicity we currently assume that the contributors are also the copyright holders, although this may not always be the case.
* A *contributors* section listing the contributors to the project.

>   Ideally the README and other parts of the documentation refer to this section.

Reviewer: ```suggestion
  Ideally, the README and other parts of the documentation refer to this section.
```

>   Contributors are all the people that have in some way contributed ad shaped the *work* of the project.

Reviewer: ```suggestion
  Contributors are all the people who have contributed to and shaped the *work* of the project in some way.
```

  This does not only include code contributions, but also design and architecture decisions, documentation, data curation, etc. .

>   Contributors should be identifiable, such that they can be contacted.

Reviewer: ```suggestion
  Contributors should be identifiable, so that they can be contacted.
```

  Common ways to list contributors are to list them with their full name and email address or their GitHub username.
* (Recommended) A *contributing* section that describes how to contribute to the project.
  From the licensing perspective, this section should include a reminder to add the license information to new files and modified files.

>   It should also make it clear that any contributions to the project are licensed under the same license as the project itself.

Reviewer: ```suggestion
  It should also be made clear that all contributions to the project are licensed under the same license as the project itself.
```

> 

Reviewer: ```suggestion

#### Soft fork specific guidelines

```

Author: I think that's already in there

Reviewer: I suggest adding a header.

> If you have a soft fork, make sure you to adjust the following parts in the documentation:

Reviewer: ```suggestion
If you have a soft fork, here are some additional recommendations:
```

  * Change the repository URL in: `html_theme_options:repository_url`.
* Reflect the content of your README in `doc/index.rst`.
  You can keep part of the original upstream text as background information if this is relevant.
* Provide an updated version of the installation guidelines in `doc/installation.rst`.
  This expands on the [README](#readme) summary.
* Provide an extensive description of the license strategy for both your repository and the upstream one in `doc/licenses.rst`.
  This should expand on the [README](#readme) summary (e.g. [Licenses](https://open-tyndp.readthedocs.io/en/latest/licenses.html) from Open-TYNDP).
* Give clear guidelines on how to contribute to and get support from this project and its upstream communities in `doc/contributing.rst` and `doc/support.rst`.
* List the people who have contributed to the project and had a significant impact on the repository in `doc/contributors.rst`.
  Link to the list of those who contributed to the original work.
  This is where you should describe the group that the SPDX license header will reference (i.e., "Contributors to [CONTRIBUTORS]") (see also [REUSE tool](#reuse-tool)).

>   Clearly state that this project builds on the work of other communities from which the soft fork was created (e.g., [Contributors](https://open-tyndp.readthedocs.io/en/latest/contributors.html) from Open-TYNDP).

Reviewer 2: ```suggestion
  Clearly state that this project builds on the work of other communities from which the soft fork was created (e.g., see the [Contributors](https://open-tyndp.readthedocs.io/en/latest/contributors.html) page from Open-TYNDP).
```
Otherwise it sounds like the upstream in this example is Open-TYNDP, which is confusing because it's the fork in many examples above.

  If the data is included in the repository, it must be covered by the licensing scheme (see [REUSE tool](#comply-with-reuse)).
  Otherwise, it needs to be listed in the relevant file in the `doc/data_sources.rst`.

### GitHub

Below are additional suggestions for configuring your GitHub repository:
* Ensure that the `About` section of your GitHub page is consistent with your README.
  This includes amending the description, the website and the [topics](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/classifying-your-repository-with-topics).
* Use a [citation file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-citation-files), such as `CITATION.cff`, to let other users know how to cite the project.
  This file is used for user convenience only and has no legal value.

>   If you use a soft fork, use the `references` field of this file to refer to the original work, see e.g. [`CITATION.cff`](https://github.com/open-energy-transition/open-tyndp/blob/master/CITATION.cff) from Open-TYNDP.

Reviewer: ```suggestion
  If you use a soft fork, use the `references` field of this file to refer to the original work. For example, see [`CITATION.cff`](https://github.com/open-energy-transition/open-tyndp/blob/master/CITATION.cff) from Open-TYNDP.
```

* Create issue and pull request templates or update the templates that you have forked in case of a soft fork, to better guide user contributions.

>   E.g. `.github/ISSUE_TEMPLATE/bug_report.yaml` and `.github/pull_request_template.md`.

Reviewer: ```suggestion
  For example, `.github/ISSUE_TEMPLATE/bug_report.yaml` and `.github/pull_request_template.md`.
```

  Add a guideline to specify what issue to report in which repository, e.g., [`bug_report.yaml`](https://github.com/open-energy-transition/open-tyndp/blob/master/.github/ISSUE_TEMPLATE/bug_report.yaml?plain=1#L19-L26) from Open-TYNDP.

## REUSE tool

> The [REUSE tool](https://reuse.software/) from the Free Software Foundation Europe simplifies the licensing stuff significantly.

Reviewer: ```suggestion
The [REUSE tool](https://reuse.software/) from the Free Software Foundation Europe simplifies the licensing significantly.
```

It helps you to ensure that your project is compliant with the [REUSE guidelines](https://reuse.software/spec/) and uses the [SPDX license identifiers](https://spdx.org/licenses/) to identify the licenses used in your project.
You can install the tool using their [installation instructions](https://reuse.readthedocs.io/en/stable/readme.html#install).
It is very good.

> We try to use it in all our projects where possible.

Reviewer: ```suggestion
It is a good tool, and we suggest using it consistently across all our projects.
```

Author: Just wanted to have some fun :)

The tool helps to record the licensing information in the repository in three standardised ways:

> 1. (Preferred) when possible the tool helps to add SPDX headers to the top of each file in the repository.

Reviewer: ```suggestion
1. (Preferred) When possible the tool helps to add SPDX headers to the top of each file in the repository.
```

>    When possible, this approach should be used.

Reviewer: ```suggestion
```

   This modification does not change the behaviour of your files.
2. (Fallback) if the file type does not support SPDX headers, the licensing information can be recorded in the `REUSE.toml` file located at the root of the repository.
   This file records the licensing information for all files that can not have an SPDX header, e.g. binary files and images.

> 3. (Disencouraged) Another way to record the licensing information is to add `.license` file for each file that cannot be annotated.

Reviewer 2: ```suggestion
3. (Discouraged) Another way to record the licensing information is to add a `.license` file for each file that cannot be annotated.
```

> 2. (Fallback) if the file type does not support SPDX headers, the licensing information can be recorded in the `REUSE.toml` file in the root of the repository.

Reviewer: ```suggestion
2. (Fallback) if the file type does not support SPDX headers, the licensing information can be recorded in the `REUSE.toml` file located at the root of the repository.
```

3. (Disencouraged) Another way to record the licensing information is to add `.license` file for each file that cannot be annotated.
   When using the second approach with `REUSE.toml`, this approach is not needed.

>    It is discouraged by us because it is not as easy to maintain as the first two approaches and creates a lot of additional files in the repository.

Reviewer: ```suggestion
   This approach is not recommended because it is more difficult to maintain than the two first approaches. It also creates a lot of additional files in the repository.
```

The SPDX headers added by the REUSE tool using `reuse annotate` usually consist of two lines:

```python
# SPDX-FileCopyrightText: ...
# SPDX-License-Identifier: ...
```

The first line contains the copyright holder, which is usually the contributors to the project, e.g. `# SPDX-FileCopyrightText: Contributors to Open-TYNDP`.

> In this line [we do not include the copyright year](https://reuse.software/faq/#years-copyright); although it is possible to include it, it is not needed and just adds another confusing element to the header that needs to be maintained.

Reviewer: ```suggestion
In this line, [we do not include the copyright year](https://reuse.software/faq/#years-copyright). Although it is possible, it is not necessary and adds another element to keep updated.
```

The second line contains the standardised SPDX license identifier for the license used in the file, e.g. `SPDX-License-Identifier: MIT` for the MIT license.

> Another optional line is to include the contributors to a file that do not hold the copyright using the `SPDX-FileContributorText` identifier.

Reviewer: ```suggestion
Another optional line is to include the contributors to a file who do not hold the copyright using the `SPDX-FileContributorText` identifier.
```

> In our case we assume that the contributors are also the copyright holders, so we do not use this line.

Reviewer: ```suggestion
In our case, we assume that the contributors are also the copyright holders, so we do not use this line.
```

### Adding a license to a fresh file

> A fresh file without any license information does not require special considerations.

Reviewer: ```suggestion
A fresh file with no license information does not require any special considerations.
```

> For the annotation you need to know which license you want to use for the file, see [OpenSource](OpenSource) for a suggestion on licenses to use.

Reviewer: ```suggestion
To add an annotation, you need to know which license you want to use for the file. See [OpenSource](OpenSource) for suggestions on licenses to use.
```

You can also check the [PyPSA-Eur repository](https://github.com/pypsa/pypsa-eur) and the [Open-TYNDP repository](https://github.com/open-energy-transition/open-tyndp) for examples.

:::info

>     To add the license information to a fresh file without any license information, run:

Reviewer: ```suggestion
    To add the license information to a fresh file with no license information, run:
```

    ```bash
    reuse annotate --license=<license identifier, e.g. mit or cc-by-4.0> --exclude-year --copyright="Contributors to [CONTRIBUTORS] <[REPOSITORY]>" <file or list of files>
    ```

:::

`reuse` does not automatically create entries for the `REUSE.toml` file, this must be done manually.
Since it is the fallback solution, this shouldn't be necessary for most files.

> Here is an example of what a `REUSE.toml` file looks like this:

Reviewer 2: ```suggestion
Here is an example of what a `REUSE.toml` file looks like:
```

> `reuse` does not automatically create entries for the `REUSE.toml` file, this needs to be done manually.

Reviewer: ```suggestion
`reuse` does not automatically create entries for the `REUSE.toml` file, this must be done manually.
```

> An example `REUSE.toml` looks like this:

Reviewer: ```suggestion
Here is an example of what a `REUSE.toml` file looks like this:
```

> ```toml
> version = 1
> 
> [[annotations]]
> path = [
>     "tests/*.py",
>     "MANIFEST",
>     ".github/workflows/*.yml",
> ]
> SPDX-FileCopyrightText = "2019 Translation Company"
> SPDX-License-Identifier = "MIT"
> 
> [[annotations]]
> path = "tests/resources/**"
> SPDX-FileCopyrightText = "2019 Jane Doe"
> SPDX-License-Identifier = "CC0-1.0"
> ```

Reviewer: ```suggestion
```toml
version = 1

[[annotations]]
path = [
    "tests/*.py",
    "MANIFEST",
    ".github/workflows/*.yml",
]
SPDX-FileCopyrightText = "Translation Company"
SPDX-License-Identifier = "MIT"

[[annotations]]
path = "tests/resources/**"
SPDX-FileCopyrightText = "Jane Doe"
SPDX-License-Identifier = "CC0-1.0"
```
```

> For more on the `REUSE.toml` file, see the [REUSE documentation](https://reuse.software/spec-3.2/) and the Open-TYNDP repository.

Reviewer: ```suggestion
For more on the `REUSE.toml` file, see the [REUSE documentation](https://reuse.software/spec-3.2/#reusetoml) and the [Open-TYNDP](https://github.com/open-energy-transition/open-tyndp/blob/master/REUSE.toml) repository.
```

### Modifying the `REUSE.toml` file

In a soft fork with an existing `REUSE.toml` file, make the following adjustments:
* Add a new package supplier for the project using `SPDX-PackageSupplier`.
* Transform the original package supplier field `SPDX-PackageSupplier` to `SPDX-PackageOriginator`.
* Adjust the package location `SPDX-PackageDownloadLocation` to point to the right repository.

### Adding licenses to an existing file

If you have an existing file that already contains a license header, you can also use the `reuse annotate` command to add the license information.

> Note that you may only claim copyright over the file if you have made contributions to the file.

Reviewer: ```suggestion
Note that you may only claim copyright over a file if you have contributed to it.
```

> Simply forking a repository does not give you the right to add a copyright header to the files in the repository.

Reviewer: ```suggestion
Simply forking a repository does not give you the right to add a copyright header to its files.
```

> If the license remains the same for the modifications, you can add the additional copyright holders to the top of the file with an additional SPDX header.

Reviewer: ```suggestion
If the license remains the same for the modifications, you can add the additional copyright holders at the top of the file by adding an additional SPDX header.
```

:::info

    To add the license information to an existing file with a license header that keeps the same license, run:

    ```bash
    reuse annotate --exclude-year --merge-copyrights --skip-unrecognised --copyright="Contributors to [CONTRIBUTORS] <[REPOSITORY]>"
    ```

:::

> For example this would turn the following header using the `reuse annotate` command:

Reviewer: ```suggestion
For example, running this `reuse annotate` command:
```

```bash
reuse annotate --exclude-year --merge-copyrights --skip-unrecognised --copyright="Contributors to Open-TYNDP <https://github.com/open-energy-transition/open-tyndp>"
```

on a file with the following header:

```python
# SPDX-FileCopyrightText: Contributors to PyPSA-Eur <https://github.com/pypsa/pypsa-eur>
#
# SPDX-License-Identifier: MIT
```

> The header becomes:

Reviewer: ```suggestion
gives this results:
```

```python
# SPDX-FileCopyrightText: Contributors to Open-TYNDP <https://github.com/open-energy-transition/open-tyndp>
# SPDX-FileCopyrightText: Contributors to PyPSA-Eur <https://github.com/pypsa/pypsa-eur>
#
# SPDX-License-Identifier: MIT
```

> If the licensing information for the file is stored in the `REUSE.toml`, you need to update the `REUSE.toml` file as well.

Reviewer: ```suggestion
If the licensing information for the file is stored in the `REUSE.toml`, you need to update that file instead.
```

It is enough to add another `SPDX-FileCopyrightText` item to the list for each `[[annotations]]` entry where this project brings additional changes.
In the example above, change the `REUSE.toml` file from this:

```toml

> [[annotations]]
> path = ["my_awesome_code.py"]
> SPDX-FileCopyrightText = [
>     "The Open-TYNDP Authors",
>     "The PyPSA-Eur Authors",
> ]
> SPDX-License-Identifier = "MIT"
> ```

Reviewer: ```suggestion
```toml
[[annotations]]
path = ["my_awesome_code.py"]
SPDX-FileCopyrightText = [
    "Translation Company",
]
SPDX-License-Identifier = "MIT"
```
```

```toml
[[annotations]]
path = ["my_awesome_code.py"]
SPDX-FileCopyrightText = [
    "Another Company",
    "Translation Company",
]
SPDX-License-Identifier = "MIT"
```

### Mixed licenses

> If you want to use a different license for your contributions to a fork, you first need to validate that the license you plan on using is compatible with the original license.

Reviewer 2: Do we have any resources we can link to where people can check such license compatibility?

Author: I've added some

> ```toml
> [[annotations]]
> path = ["my_awesome_code.py"]
> SPDX-FileCopyrightText = [
>     "The Open-TYNDP Authors",
>     "The PyPSA-Eur Authors",
> ]
> SPDX-License-Identifier = "MIT"
> ```

Reviewer: ```suggestion
```toml
[[annotations]]
path = ["my_awesome_code.py"]
SPDX-FileCopyrightText = [
    "Another Company",
    "Translation Company",
]
SPDX-License-Identifier = "MIT"
```
```

If you want to use a different license for your contributions to a fork, you first need to validate that the license you plan on using is compatible with the original license.
If this is the case, you need to add an `SPDX-FileCopyrightText` in a similar way to that used for [Identical licenses](#identical-licenses).
You also need to modify the `SPDX-License-Identifier` to also mention your new license (e.g. `SPDX-License-Identifier: MIT` to `SPDX-License-Identifier: MIT AND Apache-2.0`).
Finally, those changes need to be reflected in the `REUSE.toml` too.

> Note that this should usually not be necessary and be avoided, because of the additional complexity it adds to the project.

Reviewer: ```suggestion
Note that this is usually unnecessary and should be avoided, due to the additional complexity it adds to the project.
```

### Downloading licenses

> If you add a license to a new project or add a new license to an existing project, you need to include the license text in the repository.

Reviewer: ```suggestion
If you add a license to a new project or add a new license to an existing project, the license text must be included in the repository.
```

The REUSE tool simplifies this process immensely.

:::info
  
    To download all licenses used in your project that are missing from the `LICENSES/` directory, run:
    ```bash
    reuse download --all
    ```

:::

### Automated checks for compliance

:::info

>     Add the [`reuse` pre-commit hook](https://reuse.readthedocs.io/en/stable/readme.html#run-as-pre-commit-hook) to check your repository automatically for compliance with REUSE.

Reviewer: ```suggestion
    Add the [`reuse` pre-commit hook](https://reuse.readthedocs.io/en/stable/readme.html#run-as-pre-commit-hook) to automatically verify that your repository complies with REUSE.
```

:::

> This is the best way to ensure your repository is compliant with REUSE.

Reviewer: ```suggestion
This is the best way to ensure that your repository complies with REUSE.
```

> The hook that automatically runs `reuse lint` on every commit and will help keep your repository compliant with the REUSE guidelines.

Reviewer: ```suggestion
The hook automatically runs `reuse lint` on every commit and will help keep your repository compliant with the REUSE guidelines.
```

> It will fail if your repository is not compliant, e.g. if a new file that you have added to the repository does not have a license attached to it, or a license file is missing.

Reviewer: ```suggestion
It will fail if your repository is not compliant, e.g. if a newly added file does not have a license attached to it, or if a license file is missing.
```

## Contributions and PR templates

:::info

    Add an item to your pull request template to remind contributors to add the license information to their pull requests.

:::

It is easy to forget about the need to add licensing information to your pull request.
An additional entry can help remind contributors to add the license information to their pull request:

```md
- [ ] Licensing information to new or modified files has been added (SPDX header, REUSE.toml, etc.)
- [ ] I have added myself to the list of contributors in `doc/contributors.rst` (if applicable)
```

Adding the correct license attribution is not automatic.

> In principle it is possible to automate the step of adding license information, but it is not recommended.

Reviewer: ```suggestion
In principle, it is possible to automate the addition of license information, but this is not recommended.
```

If you remember to use tools like `reuse annotate` as described above, it becomes very easy and quick to add the license information to your files.

> See for example the [Open-TYNDP repository](https://github.com/open-energy-transition/open-tyndp/blob/6ab3be3704eb84121e8a5b5bfdf4cc45390571b5/.github/pull_request_template.md?plain=1#L26)

Reviewer: ```suggestion
See for example the [Open-TYNDP PR template](https://github.com/open-energy-transition/open-tyndp/blob/6ab3be3704eb84121e8a5b5bfdf4cc45390571b5/.github/pull_request_template.md?plain=1#L26).
```

## Contributing to other open source projects

When you contribute to other open projects, e.g. any open source or data project, some legal considerations come into play in the background.

> Projects usually require you to agree to their chosen license, e.g. the [MIT License](https://opensource.org/license/mit/) for [PyPSA](https://github.com/pypsa/pypsa).

Reviewer 2: Do we have any recommendation / policy for whether we should do the same for fresh OET-created open-source repositories?

Author: The moment we put a license on the project and someone contributes to it, they implicitly already agree. We could, maybe should, make it more explicit.
We're currently looking into this for the PyPSA repo, could follow the lead from there.

E.g. have a note in the PR template and/or a link to a very short DCO (Developer Certificate of Origin), basically saying "I have the rights to contribute the code and I contribute it under the repos license at the time of contribution).

Author: Reading into this a bit more, I'd say we are simpler off (less friction) by not having a CLA or DCO.
I think we generally assume inbound=outbound for the licenses.

Interestingly enough, GitHub has made this the plattform default in their ToS a while back: https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license

My takeaway would be to maybe modify the PR template to say something like:

> By contributing I agree to make your contributions available under the repositories MIT license.

Reviewer 2: Minor suggestion:
> By contributing I agree to make my contributions available under the repository's MIT license.

Author: Added, including a small paragraph on CLA/DCO

> When you contribute to other open projects, e.g. to an *upstream* project as described earlier, or any other open source or open data project, some legal stuff is also happening in the background.

Reviewer: ```suggestion
When you contribute to other open projects, e.g. any open source or data project, some legal considerations come into play in the background.
```

> Projects generally require you to agree with the license they have chosen for their project, e.g. the [MIT License](https://opensource.org/license/mit/) for [PyPSA](https://github.com/pypsa/pypsa).

Reviewer: ```suggestion
Projects usually require you to agree to their chosen license, e.g. the [MIT License](https://opensource.org/license/mit/) for [PyPSA](https://github.com/pypsa/pypsa).
```

Some open source projects have detailed contribution guidelines that clarify the licensing of contributions. You should read these carefully before contributing.

At OET, we have decided to grant you the full copyright of your contributions to other open source projects.
In pratice, this means that you can contribute to other open source projects freely, without requiring any additional permission from OET.

> In summary this means that....

Reviewer 2: ```suggestion
In summary this means that:
```
Otherwise it looked like an unfinished sentence/paragraph to me. I imagine the summary being referred to is the info box below?

Author: Indeed!

> Some open source projects have longer contribution guidelines that clarify the licensing of contributions in more detail, you should read those carefully before contributing.

Reviewer: ```suggestion
Some open source projects have detailed contribution guidelines that clarify the licensing of contributions. You should read these carefully before contributing.
```

> At OET we have decided that you are granted the full copyright for contributions that you make to other open source projects.

Reviewer: ```suggestion
At OET, we have decided to grant you the full copyright of your contributions to other open source projects.
```

> Effectively this means that you can freely contribute to other open source projects without requiring any additional permission from OET.

Reviewer: ```suggestion
In pratice, this means that you can contribute to other open source projects freely, without requiring any additional permission from OET.
```

---
sidebar_label: "Copyright and attribution"
---

import MaintainerTag from "@site/src/components/MaintainerTag";
import ControlledDocBanner from "@site/src/components/ControlledDocBanner";

<MaintainerTag maintainerEmails={["thomas.gilon@openenergytransition.org", "johannes.hampp@openenergytransition.org"]} />

# Copyright and attribution

This page describes how we handle copyright notices and attribution in our work, our own repositories and soft forks of other projects.

It provides generic guidelines for licensing and attributing both the original work and the work done in a specific project.
This page builds on the [Soft fork strategy](SoftForkStrategy) and covers projects that are forked from another project (referred to as `upstream` below), which is not owned by Open Energy Transition (OET).

A soft fork extends the original work in various ways, including custom code, documentation, data, and configuration.
This page assumes you already have a fully functional soft fork from another project.

> For illustrative purposes, this page uses `PyPSA-Eur` as the upstream project and `Open-TYNDP` as the soft-fork project.

Reviewer: ```suggestion
For illustrative purposes, this page uses [`PyPSA-Eur`](https://github.com/pypsa/pypsa-eur) as the upstream project and [`Open-TYNDP`](https://github.com/open-energy-transition/open-tyndp) as the soft-fork project.
```

It helps you to ensure that your project is compliant with the [REUSE guidelines](https://reuse.software/spec/) and uses the [SPDX license identifiers](https://spdx.org/licenses/) to identify the licenses used in your project.
You can install the tool using their [installation instructions](https://reuse.readthedocs.io/en/stable/readme.html#install).
It is a good tool, and we suggest using it consistently across all our projects.

The tool helps to record the licensing information in the repository in three standardised ways:

1. (Preferred) When possible the tool helps to add SPDX headers to the top of each file in the repository.
   `reuse` automatically detects the file type and adds the correct SPDX header to the top of the file.
   This modification does not change the behaviour of your files.
2. (Fallback) if the file type does not support SPDX headers, the licensing information can be recorded in the `REUSE.toml` file located at the root of the repository.
   This file records the licensing information for all files that can not have an SPDX header, e.g. binary files and images.

> 3. (Disencouraged) Another way to record the licensing information is to add `.license` file for each file that cannot be annotated.

Reviewer: ```suggestion
3. (Discouraged) Another way to record the licensing information is to add `.license` file for each file that cannot be annotated.
```

GitHub also explicitly assumes this principle for all repositories on its platform [by default](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#6-contributions-under-repository-license).
We do not use any other Contributor License Agreement (CLA) or Developer Certificate of Origin (DCO) for our projects, to keep things simple, avoid friction and encourage contributions.

## Contributing to other open source projects

When you contribute to other open projects, e.g. any open source or data project, some legal considerations come into play in the background.

Projects usually require you to agree to their chosen license, e.g. the [MIT License](https://opensource.org/license/mit/) for [PyPSA](https://github.com/pypsa/pypsa).
Some open source projects have detailed contribution guidelines that clarify the licensing of contributions. You should read these carefully before contributing.

At OET, we have decided to grant you the full copyright of your contributions to other open source projects.

> In pratice, this means that you can contribute to other open source projects freely, without requiring any additional permission from OET.

Reviewer: ```suggestion
In practice, this means that you can contribute to other open source projects freely, without requiring any additional permission from OET.
```

An additional entry can help remind contributors to add the license information to their pull request:

```md
- [ ] Licensing information to new or modified files has been added (SPDX header, REUSE.toml, etc.).

> - By contributing I agree to make my contributions available under the repository's MIT license.

Reviewer: ```suggestion
- [ ] By contributing I agree to make my contributions available under the repository's MIT license.
```
